### DISCUSSION TITLE
Docker versioning not enough for me (tags like bitnami's wordpress tags)

### DISCUSSION THREAD

**Discussion by Jonas18175** [renovatebot/renovate#12340] [Request Help] [answered]
Hello community,

i have questions about the versioning for (docker).

I use for some own images a docker tag like bitnami's wordpress - for example: `5.8.1-debian-10-r41`
The problem is that it updates only if the part before the r41 changes, but not on the single releases, but i need it to autoupdate the versions with the releases. I build for that task followed regex which is working, but it will fail when the "debian-10" part changed to "debian-11" for example.

 I know that renovate supports a <build> but i think it is not enough for that task.
My tried regex is: `^(?<major>\\d+)\\.(?<minor>\\d+\\.?\\d*?)-?(?<compatibility>.*)-r(?<patch>\\d+)?`
It works in moment, but i think it will fail in future.

I like to update when:
- 5.8.1 changes to a higher number
- debian-10 changes to debian-11 or alpine3.14.2 changes to alpine3.14.3
- the release (build number) changes to a higher value

What is the best solution to solve this task?

**Comment by viceice** [comment]
i use this 😉 

```json
    {
      "description": "Use custom regex versioning for bitnami images",
      "matchPackagePrefixes": ["gcr.io/bitnami-containers/", "bitnami/"],
      "matchDatasources": ["docker"],
      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(:?-(?<compatibility>.*-r)(?<build>\\d+))?$"
    },
```

**Reply by Jonas18175** [reply]
Thx - but i think it will not update when debian-10 changes to debian-11 - but i think it should work when i replace the compatibility with ".*" - so i can ignore that part. Or is that wrong? 

So the resulting regex should be this:
`^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(:?-.*-r(?<build>\\d+))?$`

**Reply by viceice** [reply]
This can cause image flipping, you should at least match `debian-` of `alpine` as compatibillity.

Maybe a better way is to try a regex manager which matches the version as compatibillity and the image and release versions to update.
so it should be somehting like:
```
depName: bitnami/wordpress/debian
lookupName: bitnami/wordpress
currentValue: 5.8.1-debian-10-r41
versioning: ^(?<compatibility>.*))-debian-(?<major>\\d+)-r(?<patch>\\d+)$
```

**Reply by Jonas18175** [reply]
But i think it will not update when the application version increases (5.8.1) - i tried to build a regex for the compatibility without the versionsnumber: `regex:^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-(?<compatibility>.*)(?:-)(?:\\d+\\.?\\d+\\.?(?:\\*|\\d+))-r(?<build>\\d+)$` 
But it seems that this is not working - it autoclosing my latest pull request ...
But in regex tester it look like that this stuff should working: https://regex101.com/r/HNqusI/1

EDIT:
My current renovate config looks like this:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "bumpVersion" : "patch",
  "packageRules": [
    {
      "matchUpdateTypes": ["major", "minor", "patch", "pin", "digest"],
      "automerge": false
    },
    {
      "description": "Use custom regex versioning for bitnami images",
      "matchDatasources": ["docker"],
      "matchPackageNames" : ["docker.io/bitnami/wordpress-nginx"],
      "versionScheme": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-(?<compatibility>.*)(?:-)\\d+\\.?\\d+\\.?\\d+-r(?<build>\\d+)$"
    }
  ]
}
```

**Reply by Jonas18175** [reply]
Has anybody a idea why my regex is not working inside renovate bot?

**Reply by viceice** [reply]
Check the debug logs. 🤷‍♂️

**Reply by HonkingGoose** [reply]
You can check out our Troubleshooting guide here: https://docs.renovatebot.com/troubleshooting/

**Comment by Jonas18175** [comment, marked as answer]
Now it works. I found my issue. The one or more matches after the compatibility for a single number is not working - it should be zero or unlimited.

My final regex is: `^(?<major>\\d+)\\.?(?<minor>\\d+)?\\.?(?<patch>\\d+)?-(?<compatibility>.*)(?:-\\d*\\.?\\d*\\.?\\d*)-r(?<build>\\d+)$` 

Config:
```
{
  "description": "Use custom regex versioning for bitnami images",
  "matchPackagePrefixes": ["gcr.io/bitnami-containers/", "bitnami/", "docker.io/bitnami/"],
  "versioning": "regex:^(?<major>\\d+)\\.?(?<minor>\\d+)?\\.?(?<patch>\\d+)?-(?<compatibility>.*)(?:-\\d*\\.?\\d*\\.?\\d*)-r(?<build>\\d+)$"
}
```
